File: clients/scala/src/main/scala/komus24/debugging/DebugData.scala
package komus24.debugging

import komus24.util.StreamUtil

/**
 * Data for debug rendering
 */
sealed trait DebugData {
    /**
     * Write DebugData to output stream
     */
    def writeTo(stream: java.io.OutputStream): scala.Unit
}

object DebugData {
    /**
     * Circle
     *
     * @param pos Center
     * @param radius Radius
     * @param color Color
     */
    case class Circle(pos: komus24.model.Vec2Double, radius: Double, color: komus24.debugging.Color) extends DebugData {
        /**
         * Write Circle to output stream
         */
        override def writeTo(stream: java.io.OutputStream): scala.Unit = {
            StreamUtil.writeInt(stream, Circle.TAG)
            pos.writeTo(stream)
            StreamUtil.writeDouble(stream, radius)
            color.writeTo(stream)
        }
    
        /**
         * Get string representation of Circle
         */
        override def toString(): String = {
            var stringBuilder = new StringBuilder("Circle { ")
            stringBuilder.append("pos: ")
            stringBuilder.append(pos)
            stringBuilder.append(", ")
            stringBuilder.append("radius: ")
            stringBuilder.append(radius)
            stringBuilder.append(", ")
            stringBuilder.append("color: ")
            stringBuilder.append(color)
            stringBuilder.append(" }")
            stringBuilder.toString()
        }
    }
    
    object Circle {
        val TAG: Int = 0
    
        /**
         * Read Circle from input stream
         */
        def readFrom(stream: java.io.InputStream): Circle = Circle(
            komus24.model.Vec2Double.readFrom(stream),
            StreamUtil.readDouble(stream),
            komus24.debugging.Color.readFrom(stream)
        )
    }

    /**
     * Line (segment)
     *
     * @param point1 First end
     * @param point2 Other end
     * @param width Thickness
     * @param color Color
     */
    case class Line(point1: komus24.model.Vec2Double, point2: komus24.model.Vec2Double, width: Double, color: komus24.debugging.Color) extends DebugData {
        /**
         * Write Line to output stream
         */
        override def writeTo(stream: java.io.OutputStream): scala.Unit = {
            StreamUtil.writeInt(stream, Line.TAG)
            point1.writeTo(stream)
            point2.writeTo(stream)
            StreamUtil.writeDouble(stream, width)
            color.writeTo(stream)
        }
    
        /**
         * Get string representation of Line
         */
        override def toString(): String = {
            var stringBuilder = new StringBuilder("Line { ")
            stringBuilder.append("point1: ")
            stringBuilder.append(point1)
            stringBuilder.append(", ")
            stringBuilder.append("point2: ")
            stringBuilder.append(point2)
            stringBuilder.append(", ")
            stringBuilder.append("width: ")
            stringBuilder.append(width)
            stringBuilder.append(", ")
            stringBuilder.append("color: ")
            stringBuilder.append(color)
            stringBuilder.append(" }")
            stringBuilder.toString()
        }
    }
    
    object Line {
        val TAG: Int = 1
    
        /**
         * Read Line from input stream
         */
        def readFrom(stream: java.io.InputStream): Line = Line(
            komus24.model.Vec2Double.readFrom(stream),
            komus24.model.Vec2Double.readFrom(stream),
            StreamUtil.readDouble(stream),
            komus24.debugging.Color.readFrom(stream)
        )
    }

    /**
     * Rectangle
     *
     * @param corner1 One of the corners
     * @param corner2 Opposite corner
     * @param color Color
     */
    case class Rect(corner1: komus24.model.Vec2Double, corner2: komus24.model.Vec2Double, color: komus24.debugging.Color) extends DebugData {
        /**
         * Write Rect to output stream
         */
        override def writeTo(stream: java.io.OutputStream): scala.Unit = {
            StreamUtil.writeInt(stream, Rect.TAG)
            corner1.writeTo(stream)
            corner2.writeTo(stream)
            color.writeTo(stream)
        }
    
        /**
         * Get string representation of Rect
         */
        override def toString(): String = {
            var stringBuilder = new StringBuilder("Rect { ")
            stringBuilder.append("corner1: ")
            stringBuilder.append(corner1)
            stringBuilder.append(", ")
            stringBuilder.append("corner2: ")
            stringBuilder.append(corner2)
            stringBuilder.append(", ")
            stringBuilder.append("color: ")
            stringBuilder.append(color)
            stringBuilder.append(" }")
            stringBuilder.toString()
        }
    }
    
    object Rect {
        val TAG: Int = 2
    
        /**
         * Read Rect from input stream
         */
        def readFrom(stream: java.io.InputStream): Rect = Rect(
            komus24.model.Vec2Double.readFrom(stream),
            komus24.model.Vec2Double.readFrom(stream),
            komus24.debugging.Color.readFrom(stream)
        )
    }

    /**
     * Text
     *
     * @param text Text to draw
     * @param pos Position
     * @param size Font size
     * @param align Alignment (0 - left, 0.5 - center, 1 - right)
     * @param color Color
     */
    case class Text(text: String, pos: komus24.model.Vec2Double, size: Double, align: Double, color: komus24.debugging.Color) extends DebugData {
        /**
         * Write Text to output stream
         */
        override def writeTo(stream: java.io.OutputStream): scala.Unit = {
            StreamUtil.writeInt(stream, Text.TAG)
            StreamUtil.writeString(stream, text)
            pos.writeTo(stream)
            StreamUtil.writeDouble(stream, size)
            StreamUtil.writeDouble(stream, align)
            color.writeTo(stream)
        }
    
        /**
         * Get string representation of Text
         */
        override def toString(): String = {
            var stringBuilder = new StringBuilder("Text { ")
            stringBuilder.append("text: ")
            stringBuilder.append('"' + text + '"')
            stringBuilder.append(", ")
            stringBuilder.append("pos: ")
            stringBuilder.append(pos)
            stringBuilder.append(", ")
            stringBuilder.append("size: ")
            stringBuilder.append(size)
            stringBuilder.append(", ")
            stringBuilder.append("align: ")
            stringBuilder.append(align)
            stringBuilder.append(", ")
            stringBuilder.append("color: ")
            stringBuilder.append(color)
            stringBuilder.append(" }")
            stringBuilder.toString()
        }
    }
    
    object Text {
        val TAG: Int = 3
    
        /**
         * Read Text from input stream
         */
        def readFrom(stream: java.io.InputStream): Text = Text(
            StreamUtil.readString(stream),
            komus24.model.Vec2Double.readFrom(stream),
            StreamUtil.readDouble(stream),
            StreamUtil.readDouble(stream),
            komus24.debugging.Color.readFrom(stream)
        )
    }

    /**
     * Read DebugData from input stream
     */
    def readFrom(stream: java.io.InputStream): DebugData = {
        StreamUtil.readInt(stream) match {
            case Circle.TAG => Circle.readFrom(stream)
            case Line.TAG => Line.readFrom(stream)
            case Rect.TAG => Rect.readFrom(stream)
            case Text.TAG => Text.readFrom(stream)
            case _ => throw new java.io.IOException("Unexpected tag value")
        }
    }
}

File: clients/cpp_clang/debugging/DebugData.cpp
#include "DebugData.hpp"
#include <stdexcept>

namespace debugging {

Circle::Circle(model::Vec2Double pos, double radius, debugging::Color color) : pos(pos), radius(radius), color(color) { }

// Read Circle from input stream
Circle Circle::readFrom(InputStream& stream) {
    model::Vec2Double pos = model::Vec2Double::readFrom(stream);
    double radius = stream.readDouble();
    debugging::Color color = debugging::Color::readFrom(stream);
    return Circle(pos, radius, color);
}

// Write Circle to output stream
void Circle::writeTo(OutputStream& stream) const {
    pos.writeTo(stream);
    stream.write(radius);
    color.writeTo(stream);
}

// Get string representation of Circle
std::string Circle::toString() const {
    std::stringstream ss;
    ss << "Circle { ";
    ss << "pos: ";
    ss << pos.toString();
    ss << ", ";
    ss << "radius: ";
    ss << radius;
    ss << ", ";
    ss << "color: ";
    ss << color.toString();
    ss << " }";
    return ss.str();
}

Line::Line(model::Vec2Double point1, model::Vec2Double point2, double width, debugging::Color color) : point1(point1), point2(point2), width(width), color(color) { }

// Read Line from input stream
Line Line::readFrom(InputStream& stream) {
    model::Vec2Double point1 = model::Vec2Double::readFrom(stream);
    model::Vec2Double point2 = model::Vec2Double::readFrom(stream);
    double width = stream.readDouble();
    debugging::Color color = debugging::Color::readFrom(stream);
    return Line(point1, point2, width, color);
}

// Write Line to output stream
void Line::writeTo(OutputStream& stream) const {
    point1.writeTo(stream);
    point2.writeTo(stream);
    stream.write(width);
    color.writeTo(stream);
}

// Get string representation of Line
std::string Line::toString() const {
    std::stringstream ss;
    ss << "Line { ";
    ss << "point1: ";
    ss << point1.toString();
    ss << ", ";
    ss << "point2: ";
    ss << point2.toString();
    ss << ", ";
    ss << "width: ";
    ss << width;
    ss << ", ";
    ss << "color: ";
    ss << color.toString();
    ss << " }";
    return ss.str();
}

Rect::Rect(model::Vec2Double corner1, model::Vec2Double corner2, debugging::Color color) : corner1(corner1), corner2(corner2), color(color) { }

// Read Rect from input stream
Rect Rect::readFrom(InputStream& stream) {
    model::Vec2Double corner1 = model::Vec2Double::readFrom(stream);
    model::Vec2Double corner2 = model::Vec2Double::readFrom(stream);
    debugging::Color color = debugging::Color::readFrom(stream);
    return Rect(corner1, corner2, color);
}

// Write Rect to output stream
void Rect::writeTo(OutputStream& stream) const {
    corner1.writeTo(stream);
    corner2.writeTo(stream);
    color.writeTo(stream);
}

// Get string representation of Rect
std::string Rect::toString() const {
    std::stringstream ss;
    ss << "Rect { ";
    ss << "corner1: ";
    ss << corner1.toString();
    ss << ", ";
    ss << "corner2: ";
    ss << corner2.toString();
    ss << ", ";
    ss << "color: ";
    ss << color.toString();
    ss << " }";
    return ss.str();
}

Text::Text(std::string text, model::Vec2Double pos, double size, double align, debugging::Color color) : text(text), pos(pos), size(size), align(align), color(color) { }

// Read Text from input stream
Text Text::readFrom(InputStream& stream) {
    std::string text = stream.readString();
    model::Vec2Double pos = model::Vec2Double::readFrom(stream);
    double size = stream.readDouble();
    double align = stream.readDouble();
    debugging::Color color = debugging::Color::readFrom(stream);
    return Text(text, pos, size, align, color);
}

// Write Text to output stream
void Text::writeTo(OutputStream& stream) const {
    stream.write(text);
    pos.writeTo(stream);
    stream.write(size);
    stream.write(align);
    color.writeTo(stream);
}

// Get string representation of Text
std::string Text::toString() const {
    std::stringstream ss;
    ss << "Text { ";
    ss << "text: ";
    ss << '"' << text << '"';
    ss << ", ";
    ss << "pos: ";
    ss << pos.toString();
    ss << ", ";
    ss << "size: ";
    ss << size;
    ss << ", ";
    ss << "align: ";
    ss << align;
    ss << ", ";
    ss << "color: ";
    ss << color.toString();
    ss << " }";
    return ss.str();
}

    
// Read DebugData from input stream
DebugData readDebugData(InputStream& stream) {
    switch (stream.readInt()) {
    case 0:
        return Circle::readFrom(stream);
    case 1:
        return Line::readFrom(stream);
    case 2:
        return Rect::readFrom(stream);
    case 3:
        return Text::readFrom(stream);
    default:
        throw std::runtime_error("Unexpected tag value");
    }
}

// Write DebugData to output stream
void writeDebugData(const DebugData& value, OutputStream& stream) {
    std::visit([&](auto& arg) {
        using T = std::decay_t<decltype(arg)>;
        if constexpr (std::is_same_v<T, Circle>) {
            stream.write((int) 0);
        }
        if constexpr (std::is_same_v<T, Line>) {
            stream.write((int) 1);
        }
        if constexpr (std::is_same_v<T, Rect>) {
            stream.write((int) 2);
        }
        if constexpr (std::is_same_v<T, Text>) {
            stream.write((int) 3);
        }
        arg.writeTo(stream);
    }, value);
}

// Get string representation of DebugData
std::string debugDataToString(const DebugData& value) {
    return std::visit([](auto& arg) {
        return arg.toString();
    }, value);
}


}